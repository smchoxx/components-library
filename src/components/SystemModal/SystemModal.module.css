.background {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  visibility: hidden;
  background-color: transparent;
  opacity: 0;
  transition: all 0.2s ease-in-out;
}

.show {
  z-index: 10;
  visibility: visible;
  background-color: rgb(0 0 0 / 50%);
  opacity: 1;
}

.modal {
  position: relative;
  box-sizing: border-box;
  min-width: 420px;
  padding: 28px 20px;
  background-color: #fff;
  border-radius: 20px;
  box-shadow: 0 10px 40px rgb(0 0 0 / 10%);
  opacity: 0;
  transition: all 0.2s ease-in-out;
  transform: scale(0.8);
}

.showModal {
  opacity: 1;
  transform: scale(1);
}

.header {
  position: relative;
}

.title {
  margin-bottom: 10px;
  font-size: 16px;
  font-weight: bold;
}

.closeIcon {
  position: absolute;
  top: -5px;
  right: -5px;
  cursor: pointer;
}

.content {
  font-size: 14px;
}

.footer {
  display: flex;
  gap: 10px;
  justify-content: flex-end;
  margin-top: 30px;
}

.button {
  display: flex;
  align-items: center;
  padding: 14px 16px;
  font-weight: 500;
  cursor: pointer;
  border: none;
  border-radius: 100px;
}

.confirmButton {
  color: #fff;
  background-color: #000;
}

.cancelButton {
  color: #222;
  background-color: #f2f2f2;
}

@media (width <= 525px) {
  .modal {
    min-width: 80%;
  }
}
