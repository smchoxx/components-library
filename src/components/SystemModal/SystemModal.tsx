import { ReactNode, createContext, useCallback, useContext } from 'react';
import styles from './SystemModal.module.css';
import { CloseSVG } from '../../assets/svg';

// Context
interface SystemModalContextProps {
  onClose: () => void;
}

const SystemModalContext = createContext<SystemModalContextProps | null>(null);

const useSystemModal = (): SystemModalContextProps => {
  const context = useContext(SystemModalContext);
  if (!context) {
    throw new Error('SystemModal Context Error');
  }
  return context;
};

// Component
interface SystemModalProps {
  visible: boolean;
  onClose: () => void;
  children: ReactNode;
}

function SystemModal({ visible, onClose, children }: SystemModalProps): JSX.Element {
  const value = { onClose };

  return (
    <SystemModalContext.Provider value={value}>
      <div className={[styles.background, visible ? styles.show : ''].join(' ')}>
        <div className={[styles.modal, visible ? styles.showModal : ''].join(' ')}>{children}</div>
      </div>
    </SystemModalContext.Provider>
  );
}

interface HeaderProps {
  title: string;
  isCloseIcon?: boolean;
}

const Header = ({ title, isCloseIcon = true }: HeaderProps): JSX.Element => {
  const { onClose } = useSystemModal();

  return (
    <div className={styles.header}>
      <h2 className={styles.title}>{title}</h2>
      {isCloseIcon && (
        <CloseSVG title='close-icon' className={styles.closeIcon} width={20} height={20} onClick={onClose} />
      )}
    </div>
  );
};

const Content: React.FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
  return <div className={styles.content}>{children}</div>;
};

interface ButtonProps {
  text: string;
  onClick: () => void;
}

const ConfirmButton = ({ text, onClick }: ButtonProps): JSX.Element => (
  <button className={[styles.button, styles.confirmButton].join(' ')} onClick={onClick}>
    {text}
  </button>
);
const CancelButton = ({ text, onClick }: ButtonProps): JSX.Element => (
  <button className={[styles.button, styles.cancelButton].join(' ')} onClick={onClick}>
    {text}
  </button>
);

interface FooterProps {
  onConfirm?: () => void;
  onCancel?: () => void;
  type?: 'all' | 'cancel' | 'confirm';
  confirmText?: string;
  cancelText?: string;
}

const Footer = ({ type, confirmText = '확인', cancelText = '취소', onConfirm, onCancel }: FooterProps): JSX.Element => {
  const { onClose } = useSystemModal();

  const handleConfirm = useCallback(() => {
    onConfirm && onConfirm();
    onClose();
  }, [onClose, onConfirm]);

  const handleCancel = useCallback(() => {
    onCancel && onCancel();
    onClose();
  }, [onClose, onCancel]);

  const footer = () => {
    switch (type) {
      case 'all':
        return (
          <>
            <ConfirmButton text={confirmText} onClick={handleConfirm} />
            <CancelButton text={cancelText} onClick={handleCancel} />
          </>
        );

      case 'confirm':
        return <ConfirmButton text={confirmText} onClick={handleConfirm} />;

      case 'cancel':
        return <CancelButton text={cancelText} onClick={handleCancel} />;

      default:
        return (
          <>
            <ConfirmButton text={confirmText} onClick={handleConfirm} />
            <CancelButton text={cancelText} onClick={handleCancel} />
          </>
        );
    }
  };

  return <div className={styles.footer}>{footer()}</div>;
};

SystemModal.Header = Header;
SystemModal.Content = Content;
SystemModal.Footer = Footer;

export default SystemModal;
