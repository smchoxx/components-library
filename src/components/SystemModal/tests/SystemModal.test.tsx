import { render, fireEvent, screen } from '@testing-library/react';
import SystemModal from '../SystemModal';

describe('SystemModal', () => {
  const onClose = jest.fn();
  const onConfirm = jest.fn();
  const onCancel = jest.fn();

  it('시스템 모달 랜더링 - 성공', () => {
    const { asFragment } = render(
      <SystemModal visible={true} onClose={onClose}>
        시스템 모달
      </SystemModal>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('시스템 모달 랜더링 - 실패', () => {
    expect(() => render(<SystemModal.Header title='헤더' />)).toThrowError('SystemModal Context Error');
  });

  it('시스템 모달 랜더링 - visible: false', () => {
    const { asFragment } = render(
      <SystemModal visible={false} onClose={onClose}>
        시스템 모달
      </SystemModal>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('시스템 모달 랜더링 - header (닫기 아이콘 o)', () => {
    const { asFragment } = render(
      <SystemModal visible={true} onClose={onClose}>
        <SystemModal.Header title='헤더' />
      </SystemModal>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('시스템 모달 랜더링 - header (닫기 아이콘 x)', () => {
    const { asFragment } = render(
      <SystemModal visible={true} onClose={onClose}>
        <SystemModal.Header title='헤더' isCloseIcon={false} />
      </SystemModal>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('시스템 모달 랜더링 - content', () => {
    const { asFragment } = render(
      <SystemModal visible={true} onClose={onClose}>
        <SystemModal.Content>콘텐츠 영역</SystemModal.Content>
      </SystemModal>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('시스템 모달 랜더링 - footer', () => {
    it('타입: all', () => {
      const { asFragment } = render(
        <SystemModal visible={true} onClose={onClose}>
          <SystemModal.Footer
            type='all'
            confirmText='확인'
            cancelText='
          취소'
            onConfirm={onConfirm}
            onCancel={onCancel}
          />
        </SystemModal>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('타입 없을 경우', () => {
      const { asFragment } = render(
        <SystemModal visible={true} onClose={onClose}>
          <SystemModal.Footer
            confirmText='확인'
            cancelText='
          취소'
            onConfirm={onConfirm}
            onCancel={onCancel}
          />
        </SystemModal>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('타입: confirm', () => {
      const { asFragment } = render(
        <SystemModal visible={true} onClose={onClose}>
          <SystemModal.Footer type='confirm' confirmText='확인' onConfirm={onConfirm} />
        </SystemModal>
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('타입: cancel', () => {
      const { asFragment } = render(
        <SystemModal visible={true} onClose={onClose}>
          <SystemModal.Footer type='cancel' cancelText='취소' onCancel={onCancel} />
        </SystemModal>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('확인 버튼을 클릭하면 모달이 닫힌다.', () => {
    render(
      <SystemModal visible={true} onClose={onClose}>
        <SystemModal.Footer type='all' confirmText='확인' cancelText='취소' onConfirm={onConfirm} onCancel={onCancel} />
      </SystemModal>
    );

    fireEvent.click(screen.getByText('확인'));

    expect(onConfirm).toHaveBeenCalled();
    expect(onClose).toHaveBeenCalled();
  });

  it('취소 버튼을 클릭하면 모달이 닫힌다.', () => {
    render(
      <SystemModal visible={true} onClose={onClose}>
        <SystemModal.Footer type='all' confirmText='확인' cancelText='취소' onConfirm={onConfirm} onCancel={onCancel} />
      </SystemModal>
    );

    fireEvent.click(screen.getByText('취소'));

    expect(onCancel).toHaveBeenCalled();
    expect(onClose).toHaveBeenCalled();
  });
});
