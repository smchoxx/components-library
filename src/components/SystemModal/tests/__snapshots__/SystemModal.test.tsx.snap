// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemModal 시스템 모달 랜더링 - content 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="content"
      >
        콘텐츠 영역
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - footer 타입 없을 경우 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="footer"
      >
        <button
          class="button confirmButton"
        >
          확인
        </button>
        <button
          class="button cancelButton"
        >
          
          취소
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - footer 타입: all 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="footer"
      >
        <button
          class="button confirmButton"
        >
          확인
        </button>
        <button
          class="button cancelButton"
        >
          
          취소
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - footer 타입: cancel 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="footer"
      >
        <button
          class="button cancelButton"
        >
          취소
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - footer 타입: confirm 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="footer"
      >
        <button
          class="button confirmButton"
        >
          확인
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - header (닫기 아이콘 o) 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="header"
      >
        <h2
          class="title"
        >
          헤더
        </h2>
        <reactcomponent
          class="closeIcon"
          height="20"
          title="close-icon"
          width="20"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - header (닫기 아이콘 x) 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      <div
        class="header"
      >
        <h2
          class="title"
        >
          헤더
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - visible: false 1`] = `
<DocumentFragment>
  <div
    class="background "
  >
    <div
      class="modal "
    >
      시스템 모달
    </div>
  </div>
</DocumentFragment>
`;

exports[`SystemModal 시스템 모달 랜더링 - 성공 1`] = `
<DocumentFragment>
  <div
    class="background show"
  >
    <div
      class="modal showModal"
    >
      시스템 모달
    </div>
  </div>
</DocumentFragment>
`;
